sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardRecordsIterator.java

```diff
@@ -77,20 +77,28 @@ public ShardRecordsIterator(final ShardCheckpoint initialCheckpoint,
 
   private void readMoreIfNecessary() throws TransientKinesisException {
     if (data.isEmpty()) {
-      GetKinesisRecordsResult response;
+      GetKinesisRecordsResult response = fetchRecords();
+      data.addAll(filter.apply(response.getRecords(), checkpoint));
+    }
+  }
+
+  private GetKinesisRecordsResult fetchRecords() throws TransientKinesisException {
+    GetKinesisRecordsResult response = null;
+    do {
       try {
         response = kinesis.getRecords(shardIterator, checkpoint.getStreamName(),
             checkpoint.getShardId());
+        shardIterator = response.getNextShardIterator();
       } catch (ExpiredIteratorException e) {
         LOG.info("Refreshing expired iterator", e);
         shardIterator = checkpoint.getShardIterator(kinesis);
-        response = kinesis.getRecords(shardIterator, checkpoint.getStreamName(),
-            checkpoint.getShardId());
       }
-      LOG.debug("Fetched {} new records", response.getRecords().size());
-      shardIterator = response.getNextShardIterator();
-      data.addAll(filter.apply(response.getRecords(), checkpoint));
-    }
+    } while (response == null || gotEmptyResponseButIsBeforeEndOfTheStream(response));
```

Comment:
Hmm, the control flow here looks a bit convoluted. Can you refactor this into a "while(true)" loop with explicit breaks on exit conditions? Currently I'm confused by seeing "while (response == null..." whereas response is being used at line 91.

Reply:
Actually I really prefer to not use breaks in the code so that the loop is controlled in one place only. This loop is also kind of a temporary solution (or more of a workaround) for Kinesis behaviour. Kinesis can return 0 records during iteration before reaching the end of the stream while there may be some more events waiting. Current KinesisReader.advance() logic would treat that case as reaching the end of the stream and move the watermark to current timestamp.
I'm planning to implement Kinesis reading in the background threads so that's only a temporary loop that won't be needed anymore then.